### PR TITLE
`cholesky_inverse`: complex autograd, forward AD and correct tests.

### PR DESCRIPTION
**Pull request by nikitaved**
As per title.

### CODE REVIEW

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -5468,21 +5468,29 @@ def __init__(self,
...
-        a.requires_grad = requires_grad
-        out.append(SampleInput(a))
-        out.append(SampleInput(a.clone().requires_grad_(requires_grad), kwargs=dict(upper=True)))
-    return out
+    test_cases = (torch.linalg.cholesky(a, upper=False) for a in inputs)
+    for l in test_cases:
+        # generated lower-triangular samples
+        l.requires_grad = requires_grad
+        yield SampleInput(l)  # upper=False by default
+        yield SampleInput(l.clone().requires_grad_(requires_grad), kwargs=dict(upper=False))
```
Comment (soulitzer) [inline torch/testing/_internal/common_methods_invocations.py:5489]: I see that this was already the case before, but doing `requires_grad_`  here should be a no-op since the output of l.clone() should already require grad if requires_grad=True or not require grad if requires_grad=False

Same for the upper-triangular case below.

File: torch/testing/_internal/common_methods_invocations.py [hunk 2]
```diff
@@ -5468,21 +5468,29 @@ def __init__(self,
 
 
 def sample_inputs_linalg_cholesky_inverse(op_info, device, dtype, requires_grad=False, **kwargs):
-    # Generate Cholesky factors of positive-definite (non-singular) Hermitian (symmetric) matrices
```
Comment (soulitzer) [inline torch/testing/_internal/common_methods_invocations.py:5471]: Does this still apply?

File: torch/testing/_internal/common_methods_invocations.py [hunk 3]
```diff
@@ -8341,6 +8348,21 @@ def gradcheck_wrapper_triangular_input(op, *args, upper=False, idx=0, **kwargs):
...
 
+def gradcheck_wrapper_triangular_input_real_positive_diagonal(op, *args, upper=False, idx=0, **kwargs):
+    """Gradcheck wrapper for functions that take lower/upper triangular matrices
+    with real and positive diagonals, for example, cholesky-like operations.
+    """
+    arg = args[idx]
+    arg_diag = arg.diagonal(0, -2, -1)
+    arg_diag_embed = torch.diag_embed(arg_diag)
+    # new_arg = arg - diag(arg) + I
+    new_arg = arg - arg_diag_embed + torch.diag_embed(torch.ones_like(arg_diag))
```
Comment (soulitzer) [inline torch/testing/_internal/common_methods_invocations.py:8359]: Looks like the diag is always ones here? Is that okay or do we want to do random positive values?

**Review by soulitzer — APPROVED** [review]
Looks good, thanks! Just some small comments/questions.

File: torch/testing/_internal/common_methods_invocations.py (hunk 2 above)
Comment (nikitaved) [reply torch/testing/_internal/common_methods_invocations.py:5471]: Do we keep it? Actually, it would make sense to update the doc on cholesky that the diagonal of a cholesky factor is real and non-negotiable.

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 4]
```diff
@@ -1192,19 +1192,35 @@ Tensor cholesky_inverse_backward(Tensor grad, Tensor L, bool upper, Tensor inver
...
-  } else {
-    grad_L = at::zeros({1}, L.options()).expand_as(L);
   }
+
   return grad_L;
 }
 
+// If X = (L L^H)^{-1} with L lower-triangular with a real positive diagonal,
+// then dX = K^H + K, where
+// K =  L^{-H} dL^{-1} = [dL^{-1} = -L^{-1} dL L^{-1}]
```
[the remark this answers is not in the document]
Comment (nikitaved) [reply torch/csrc/autograd/FunctionsManual.cpp:1209]: Yes. These are equivalent expressions.

File: torch/testing/_internal/common_methods_invocations.py (hunk 1 above)
Comment (nikitaved) [reply torch/testing/_internal/common_methods_invocations.py:5489]: I will replace with `clone().detach()`. Does `clone` always create a leaf?

File: torch/testing/_internal/common_methods_invocations.py [hunk 5]
```diff
@@ -5468,21 +5468,30 @@ def __init__(self,
...
-        a.requires_grad = requires_grad
-        out.append(SampleInput(a))
-        out.append(SampleInput(a.clone().requires_grad_(requires_grad), kwargs=dict(upper=True)))
-    return out
+    test_cases = (torch.linalg.cholesky(a, upper=False) for a in inputs)
+    for l in test_cases:
+        # generated lower-triangular samples
+        l.requires_grad = requires_grad
+        yield SampleInput(l)  # upper=False by default
+        yield SampleInput(l.clone().detach().requires_grad_(requires_grad), kwargs=dict(upper=False))
```
Comment (nikitaved) [inline torch/testing/_internal/common_methods_invocations.py:5490]: make detach first.

File: torch/testing/_internal/common_methods_invocations.py [hunk 6]
```diff
@@ -5468,21 +5468,30 @@ def __init__(self,
...
-    return out
+    test_cases = (torch.linalg.cholesky(a, upper=False) for a in inputs)
+    for l in test_cases:
+        # generated lower-triangular samples
+        l.requires_grad = requires_grad
+        yield SampleInput(l)  # upper=False by default
+        yield SampleInput(l.detach().clone().requires_grad_(requires_grad), kwargs=dict(upper=False))
+
+        # generate upper-triangular inputs
+        u = l.detach().mT.contiguous().requires_grad_(requires_grad)
```
Comment (nikitaved) [inline torch/testing/_internal/common_methods_invocations.py:5493]: Make a clone here as there is no guarantee that `contiguous` will create a copy since `l` is fortran-contiguous, so transposing it will turn it into a contiguous tensor, which will make `u` and `l` not independent.

File: torch/testing/_internal/common_methods_invocations.py [hunk 7]
```diff
@@ -8346,6 +8354,25 @@ def gradcheck_wrapper_triangular_input(op, *args, upper=False, idx=0, **kwargs):
...
+    """Gradcheck wrapper for functions that take lower/upper triangular matrices
+    with real and positive diagonals, for example, cholesky-like operations.
+    """
+    arg = args[idx]
+    arg_diag = arg.diagonal(0, -2, -1)
+    arg_diag_embed = torch.diag_embed(arg_diag)
+    id_diag_tensor = torch.ones_like(arg_diag)
+    # if arg.requires_grad, mark this tensor with requires_grad=True,
+    # so that the diagonal grads are non-zero.
+    id_tensor = torch.diag_embed(id_diag_tensor).requires_grad_(arg.requires_grad)
```
Comment (soulitzer) [inline torch/testing/_internal/common_methods_invocations.py:8367]: The entire `new_arg` tensor should require grad as long as `arg` requires grad, so we probably shouldn't need to make this tensor `require_grad_`. Requires-gradness is specified at the tensor level, autograd doesn't have fine-grained control over which particular elements of that tensor should or shouldn't require grad.

Comment (nikitaved) [reply torch/testing/_internal/common_methods_invocations.py]: removed.
